pkg/receive/handler.go

```diff
@@ -737,24 +749,40 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
 		})
 		if err != nil {
 			level.Debug(tLogger).Log("msg", "local tsdb write failed", "err", err.Error())
-			responses <- newWriteResponse(wreqs[writeTarget].seriesIDs, errors.Wrapf(err, "store locally for endpoint %v", writeTarget.endpoint))
+			responses <- newWriteResponse(wreqs[writeTarget].seriesIDs, errors.Wrapf(err, "store locally for endpoint %v", writeTarget.endpoint), writeTarget)
+			continue
+		}
+
+		responses <- newWriteResponse(wreqs[writeTarget].seriesIDs, nil, writeTarget)
+	}
+
+	reduceWaitingFor := func() {
+		n := waitingFor.Add(-1)
+		if n == 0 {
+			close(responses)
+		}
+	}
+
+	for writeTarget := range wreqs {
+		if writeTarget.endpoint == h.options.Endpoint {
 			continue
 		}
+		waitingFor.Add(1)
+	}
 
-		responses <- newWriteResponse(wreqs[writeTarget].seriesIDs, nil)
+	if waitingFor.Load() == 0 {
+		close(responses)
```

Comment:
shouldnt we just "defer close(responses)" after declaring the channel?

Reply:
No because a goroutine might still send a response over this channel after the function ends. With `defer close(responses)` it's very likely that a goroutine will panic due to sending a message to a closed channel.